### PR TITLE
Fix PT2.0 defects: Apex torch._six, training toolkit

### PR DESCRIPTION
**Pull request by ShiboXing**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
1. fix torch._six error in apex package, added upstream DDP apex test
2. patch flaky NCCL OOM error in 8-GPU instances
3. reduce the frequency of the loss logging for easier inspection

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by ShiboXing** [comment]
SM test:
efa: 9b9fb4f
standard: 12d59b2

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py [hunk 1]
```diff
@@ -66,3 +66,35 @@ def test_pytorchddp_throughput_gpu(
...
             ecr_image, sagemaker_regions, estimator_parameter, job_name=job_name
         )
+
+@pytest.mark.processor("gpu")
+@pytest.mark.model("N/A")
+@pytest.mark.multinode(2)
+@pytest.mark.integration("pytorchddp")
+@pytest.mark.parametrize(
+    "efa_instance_type", get_efa_test_instance_type(default=["ml.p4d.24xlarge"]), indirect=True
+)
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py:76]: Why do we need to run this test using a p4d.24xlarge, and not, say a p3.8xlarge/16xlarge?

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def can_run_pytorchddp(ecr_image):
 def test_pytorchddp_throughput_gpu(
     framework_version, ecr_image, sagemaker_regions, efa_instance_type, tmpdir
 ):
-    with timeout(minutes=DEFAULT_TIMEOUT):
+    with timeout(minutes=25):
```
[the remark this answers is not in the document]
Comment (ShiboXing) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py:52]: The default is used in many other tests. We can change the timeout to specific tests to better catch hanging issue, as well as performance issues more accurately.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py (hunk 1 above)
Comment (ShiboXing) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py:77]: Changed!

File: pytorch/training/docker/2.0/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -248,9 +248,9 @@ ARG PT_TORCHDATA_URL
 
 # Install SM AWS-PyTorch (force remove will increase the docker image size drastically)
 RUN /opt/conda/bin/mamba install -y pytorch=${PYTORCH_VERSION} \
-  torchvision torchaudio torchtext cpuonly \
+  torchvision torchaudio torchtext cpuonly aws-ofi-nccl-dlc \
```
Comment (saimidu) [inline pytorch/training/docker/2.0/py3/Dockerfile.cpu:251]: Why is aws-ofi-nccl needed on a CPU image that doesn't actually have NCCL installed?

File: pytorch/training/docker/2.0/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -250,13 +250,13 @@ ARG PT_TORCHDATA_URL
 RUN /opt/conda/bin/mamba install -y pytorch=${PYTORCH_VERSION} \
   torchvision torchaudio torchtext cpuonly \
   --override-channels \
-  -c https://aws-ml-conda-pre-prod-ec2.s3.us-west-2.amazonaws.com \
+  -c https://aws-ml-conda-pre-prod-sm.s3.us-west-2.amazonaws.com \
```
Comment (ohadkatz) [inline pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: Is there a reason we are using SM in the EC2 recipe?

File: test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        actual = param.grad.data.sum().item()
+        print(
+            name
+            + ": grad.data_ptr() = {}, expected sum {}, got {}".format(
+                param.grad.data_ptr(), expected, actual
+            )
+        )
+        return expected == actual
+
+    if not info("model.a", model.module.a, 2.0):
```
Comment (ohadkatz) [inline test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py:71]: Nit: These two conditionals can be condensed into one, unless there is a specific case for each of these, in which case that should be logged. Do we want to also hard stop the test after the failed pass?

File: pytorch/training/docker/2.0/py3/Dockerfile.cpu (hunk 4 above)
Comment (saimidu) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: I think this comes from the bug fixes.

Comment (kace) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: This is strange. Can we confirmed with @roywei?

File: test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py (hunk 5 above)
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py:71]: I think it would make debugging easier to have these two function calls in separate lines of code.

By running both `module.a` and `module.b` regardless of whether the other failed will give us results from both tests, and a difference in success/failure between them both will give us more information than just one test.

Comment (ohadkatz) [reply test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py:71]: We might want to log that a certain module failed and hard fail at that point, if this were to fail after the fact it wouldn't provide any context into which model failed, and would not save any additional time.

File: pytorch/training/docker/2.0/py3/Dockerfile.cpu (hunk 4 above)
Comment (ohadkatz) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: I see its ec2 here - https://github.com/aws/deep-learning-containers/pull/2899/files#diff-bc0b77b02ee5079e588e1fd8714585bee92e4159c585a980f2f576beffbc6228L271

Comment (roywei) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: this is coming from bug fixes. Debugger patch is causing issues with SMP and we disabled it. Hence using EC2 binaries in SM DLC for now.

Comment (ohadkatz) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: @roywei I see EC2 in SM, and SM in EC2. Is that intentional?

Comment (roywei) [reply pytorch/training/docker/2.0/py3/Dockerfile.cpu:253]: sorry for the confusion, for this specific line, it should be ec2, it should all be ec2 binaries for PT 2.0 release, as we disabled debugger and Rubik patch is no longer needed

File: test/sagemaker_tests/pytorch/training/resources/distributed_operations.py [hunk 6]
```diff
@@ -106,10 +106,10 @@ def _all_reduce(rank, rows, columns):
 def _reduce(rank, rows, columns):
     dest = 0
     tensor = _get_tensor(rank, rows, columns)
-    logger.debug("Rank: {},\nTensor BEFORE reduce: {}".format(rank, tensor))
+    # logger.debug("Rank: {},\nTensor BEFORE reduce: {}".format(rank, tensor))
```
Comment (ohadkatz) [inline test/sagemaker_tests/pytorch/training/resources/distributed_operations.py:109]: NIT: Can we remove these lines instead of commenting them out?

Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/resources/distributed_operations.py:109]: I think the idea was to keep them for future use, if needed, where the debug logs for these parts can be enabled if more information is needed when debugging.

Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/resources/distributed_operations.py]: Uncommented these lines, and running tests in [afcaccf](https://github.com/aws/deep-learning-containers/pull/2899/commits/afcaccf98f024cf3e7db8fd37f80eb8da7555e4f)

**Review by ohadkatz — APPROVED** [review]
